### PR TITLE
feat(stepfunctions-tasks): add sns publish with message attributes

### PR DESCRIPTION
**Pull request by moofish32**
This PR address #4702

I am primarily solving for the use case of using SNS publish with
message attributes for sns filtering. If EventBridge integrations have a
fast follow for wait for task token features, that would solve my use
case. Because I'm solving for filtering and based on anecdotal
experience String is the primary attribute type. I've included Binary
because the pattern is the same. I have not implemented Number or
String.Array because they are not Lambda supported and I believe less
common.

I chose to attempt to solve for consumer ergonomics with:

```ts
    const task = new SnsPublish(stack, 'Publish', {
      topic,
      message: sfn.TaskInput.fromText('Publish this message'),
      messageAttributes: {
        cake: {
          type: SnsMessageAttributeType.STRING,
          value: sfn.TaskInput.fromText('chocolate'),
        },
        cakePic: {
          type: SnsMessageAttributeType.BINARY,
          value: sfn.TaskInput.fromDataAt('$.cake.pic'),
        },
      },
    });
```

This results in a `value` member on the message attribute interface.
While I think that maps best for the SNS json definition and does yield
this ugly line in the private render methods:

```
      attrs[a][`${attr.type}Value`] = attr.value.value;
```

Is there a better way?

I did not implement SQS message attributes yet. Based on my limited
knowledge the StepFunction integrations are similar(potentially
identical if we again only support Binary and String). If the answer is
one implementation, should I move the below code snippets into something
generically named MessageAttribute within the stepfunctions package
perhaps as part of base-task?

```ts
/**
 * SNS Message Attribute type enum
 */
export enum SnsMessageAttributeType {

  /**
   * String type attributes.
   *
   * @see https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html#SNSMessageAttributes.DataTypes
   */
  STRING = 'String',

  /**
   * Binary type attributes.
   *
   * @see https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html#SNSMessageAttributes.DataTypes
   */
  BINARY = 'Binary',
}

/**
 * SNS Message Attribute
 */
export interface SnsMessageAttribute {
  /**
   * MessageAttributeType is the type of data being passed.
   *
   *
   * @see https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html#SNSMessageAttributes.DataTypes
   */
  readonly type: SnsMessageAttributeType;

  /**
   * The value of the data being passed
   */
  readonly value: sfn.TaskInput;
}
```

### CODE REVIEW

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 1]
```diff
@@ -4,6 +4,44 @@ import * as sfn from '@aws-cdk/aws-stepfunctions';
 import { Construct } from 'constructs';
 import { integrationResourceArn, validatePatternSupported } from '../private/task-utils';
 
+/**
+ * SNS Message Attribute type enum
+ */
+export enum SnsMessageAttributeType {
```
Comment (moofish32) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:10]: If this is common to SNS & SQS should I rename this and move it to aws-stepfunctions base task?

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 2]
```diff
@@ -98,8 +147,30 @@ export class SnsPublish extends sfn.TaskStateBase {
...
+  private _renderAttributes(): any {
+    if (!this.props.messageAttributes) {
+      return undefined;
+    }
+    const attrs: { [key: string]: any } = {};
+    for (const a of Object.keys(this.props.messageAttributes)) {
+      const attr = this.props.messageAttributes[a];
+      attrs[a] = {};
+      attrs[a].DataType = attr.type;
+      attrs[a][`${attr.type}Value`] = attr.value.value;
```
Comment (moofish32) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:170]: I don't love the need for this function or this specific line. Is there a better solution?

**Review by moofish32 — COMMENTED** [review]
Adding in line comments similar to the summary for comment tracking.

**Comment by moofish32** [comment]
@shivlaks I think you indicated some interest in a PR here, if you have something ready feel free to push and take the lead.

**Comment by moofish32** [comment]
Ok I've realized I did not understand the value/flexibility of `sfn.TaskInput.fromObject`. I think we can simplify by using that function. Can anybody from the CDK team comment if that is a better path? I have pushed a very rough implementation example using this approach. 

If we can agree that we will only support `String` types then this solution could simplify to a structure more like:

```ts
    const pubTask = new SnsPublish(stack, 'Publish', {
      topic,
      message: sfn.TaskInput.fromText('Publish this message'),
      messageAttributes: {
        cake: 'chocolate',
        attribute2: sfn.TaskInput.fromDataAt('$.attribute2'),
      },
    });
```

I prefer the above because it's the only use case I've seen in the wild and most simple to declare.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts (hunk 1 above)
Comment (shivlaks) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:10]: I think it makes sense to model them separately

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 3]
```diff
@@ -4,6 +4,49 @@ import * as sfn from '@aws-cdk/aws-stepfunctions';
...
+}
+
+/**
+ * SNS Message Attribute
+ */
+export interface SnsMessageAttribute {
+  /**
+   * Name is the name of the message attribute
+   */
+  readonly name: string;
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:34]:
```suggestion
  /**
   * Name of the message attribute
   */
  readonly name: string;
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 4]
```diff
@@ -4,6 +4,49 @@ import * as sfn from '@aws-cdk/aws-stepfunctions';
...
+  /**
+   * MessageAttributeType is the type of data being passed.
+   *
+   *
+   * @see https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html#SNSMessageAttributes.DataTypes
+   * @default STRING and StringValue will be set
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:40]:
```suggestion
   * @default SnsMessageAttributeType.STRING
```

what does the StringValue refer to?

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 5]
```diff
@@ -23,6 +66,17 @@ export interface SnsPublishProps extends sfn.TaskStateBaseProps {
...
   readonly message: sfn.TaskInput;
 
+  /**
+   * Add message attributes when publishing.
+   *
+   * For example, these attributes can carry additional metadata and may be used
+   * for SNS Filter Subscriptions.
+   *
+   * @see https://docs.aws.amazon.com/sns/latest/api/API_Publish.html#API_Publish_RequestParameters
+   * @default no attributes are sent
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:76]:
```suggestion
   * @default - no attributes are sent
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 6]
```diff
@@ -92,14 +159,17 @@ export class SnsPublish extends sfn.TaskStateBase {
...
    */
   protected _renderTask(): any {
+
     return {
       Resource: integrationResourceArn('sns', 'publish', this.integrationPattern),
       Parameters: sfn.FieldUtils.renderObject({
         TopicArn: this.props.topic.topicArn,
         Message: this.props.message.value,
         MessageStructure: this.props.messagePerSubscriptionType ? 'json' : undefined,
+        MessageAttributes: this.messageAttributes?.value,
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:169]: can messageAttributes be expressed dynamically through a path?

i.e. is the following allowed?

```js
MessageAttributes.$: '$.somePath'
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 7]
```diff
@@ -77,6 +133,17 @@ export class SnsPublish extends sfn.TaskStateBase {
...
 
+    if (props.messageAttributes) {
+      const attributes: { [key: string]: any } = {};
+      for (const attrib of props.messageAttributes) {
+        const dataType = attrib.type || SnsMessageAttributeType.STRING;
+        attributes[attrib.name] = {
+          DataType: dataType,
+          StringValue: attrib.value.value,
+        };
+      };
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:144]: nit: would be nice to move this into a utility method to keep the constructor minimal.

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/sns/publish.test.ts [hunk 8]
```diff
@@ -38,6 +38,61 @@ describe('Publish', () => {
...
+    const topic = new sns.Topic(stack, 'Topic');
+
+    // WHEN
+    const task = new SnsPublish(stack, 'Publish', {
+      topic,
+      message: sfn.TaskInput.fromText('Publish this message'),
+      messageAttributes: [
+        {
+          name: 'cake',
+          type: SnsMessageAttributeType.STRING,
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/test/sns/publish.test.ts:53]: don't think this is needed in this test since that's the default right?

**Review by shivlaks — DISMISSED** [review]

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts (hunk 4 above)
Comment (moofish32) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:40]: StringValue is the type that must be set in the JSON States language. There are two primary types Binary/BinaryValue and String/StringValue. In addition to these in some languages the [API ](https://docs.aws.amazon.com/sns/latest/api/API_Publish.html) supports other [variations](https://docs.aws.amazon.com/sns/latest/api/API_MessageAttributeValue.html) I think based on use cases StepFunction is going to very rarely need to support binary. The most common need here based on the issue is filter policies based on strings. You can see this usage with `DataType` and `StringValue` in the examples [here](https://docs.aws.amazon.com/step-functions/latest/dg/connect-sns.html).

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/sns/publish.test.ts (hunk 8 above)
Comment (moofish32) [reply packages/@aws-cdk/aws-stepfunctions-tasks/test/sns/publish.test.ts:53]: true - we can remove it.

**Comment by moofish32** [comment]
> Ok I've realized I did not understand the value/flexibility of `sfn.TaskInput.fromObject`. I think we can simplify by using that function. Can anybody from the CDK team comment if that is a better path? I have pushed a very rough implementation example using this approach.
> 
> If we can agree that we will only support `String` types then this solution could simplify to a structure more like:
> 
> ```ts
>     const pubTask = new SnsPublish(stack, 'Publish', {
>       topic,
>       message: sfn.TaskInput.fromText('Publish this message'),
>       messageAttributes: {
>         cake: sfn.TaskInput.fromText('chocolate'),
>         attribute2: sfn.TaskInput.fromDataAt('$.attribute2'),
>       },
>     });
> ```
> 
> I prefer the above because it's the only use case I've seen in the wild and most simple to declare.
> 
> @shivlaks are you able to help here at all?

I think I still have a question here. If we agree String support is the only type we can simplify a lot of code. 

```
export interface MessageAttributes { [key: string]: string; }
// usage becomes
const pubTask = new SnsPublish(stack, 'Publish', {
      topic,
      message: sfn.TaskInput.fromText('Publish this message'),
      messageAttributes: {
        cake: 'chocolate',
        attribute2: sfn.JsonPathAt('$.attribute2'),
      },
    });
```

I think part of my confusion is where do I need to enforce they use sfn.TaskInput versus just using literal strings and JsonPath?

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 9]
```diff
@@ -77,6 +133,17 @@ export class SnsPublish extends sfn.TaskStateBase {
       }
     }
 
+    if (props.messageAttributes) {
+      const attributes: { [key: string]: any } = {};
+      for (const attrib of props.messageAttributes) {
+        const dataType = attrib.type || SnsMessageAttributeType.STRING;
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:139]: prefer ?? for nullish coalescing, but see other comments since this will probably change
```suggestion
        const dataType = attrib.type ?? SnsMessageAttributeType.STRING;
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 10]
```diff
@@ -23,6 +66,17 @@ export interface SnsPublishProps extends sfn.TaskStateBaseProps {
...
   readonly message: sfn.TaskInput;
 
+  /**
+   * Add message attributes when publishing.
+   *
+   * For example, these attributes can carry additional metadata and may be used
+   * for SNS Filter Subscriptions.
+   *
+   * @see https://docs.aws.amazon.com/sns/latest/api/API_Publish.html#API_Publish_RequestParameters
+   * @default - no attributes are sent
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:76]:
```suggestion
   * @default []
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 11]
```diff
@@ -23,6 +66,17 @@ export interface SnsPublishProps extends sfn.TaskStateBaseProps {
    */
   readonly message: sfn.TaskInput;
 
+  /**
+   * Add message attributes when publishing.
+   *
+   * For example, these attributes can carry additional metadata and may be used
+   * for SNS Filter Subscriptions.
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:73]:
```suggestion
   * These attributes carry additional metadata about the message and may be used
   * for subscription filters.
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 12]
```diff
@@ -23,6 +66,17 @@ export interface SnsPublishProps extends sfn.TaskStateBaseProps {
...
+   *
+   * @see https://docs.aws.amazon.com/sns/latest/api/API_Publish.html#API_Publish_RequestParameters
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:75]:
```suggestion
   * @see https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 13]
```diff
@@ -77,6 +133,17 @@ export class SnsPublish extends sfn.TaskStateBase {
       }
     }
 
+    if (props.messageAttributes) {
+      const attributes: { [key: string]: any } = {};
+      for (const attrib of props.messageAttributes) {
+        const dataType = attrib.type || SnsMessageAttributeType.STRING;
+        attributes[attrib.name] = {
+          DataType: dataType,
+          StringValue: attrib.value.value,
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:142]: this needs to be BinaryValue if the attribute type is Binary

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts (hunk 4 above)
Comment (BenChaimberg) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:40]: it would be really cool if we could set this automatically based on the type of the value. if the value is from the input (JsonPath.isEncodedJsonPath) or unresolved (Token.isUnresolved) then we default to string, otherwise we check if the type is a string, number, or array. if none of the above (and arrays) we can serialize the value as well

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 14]
```diff
@@ -4,6 +4,49 @@ import * as sfn from '@aws-cdk/aws-stepfunctions';
 import { Construct } from 'constructs';
 import { integrationResourceArn, validatePatternSupported } from '../private/task-utils';
 
+/**
+ * SNS Message Attribute type enum
+ */
+export enum SnsMessageAttributeType {
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:10]: need to support other types

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts (hunk 6 above)
Comment (BenChaimberg) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:169]: I'm fine with not supporting this case

**Review by BenChaimberg — CHANGES_REQUESTED** [review]

**Comment by BenChaimberg** [comment]
I think overall, I'd like to support more than just strings. The current split of only allowing task input in the attribute value is fine for me. This is a design decision that we should align on, but I think it's a realistic tradeoff as allowing too much flexibility removes all the nice type checking, serialization, etc. that we can do

**Comment by moofish32** [comment]
I’ve got some free time next week. I like the feedback I’ll make some
changes.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts (hunk 4 above)
Comment (moofish32) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts]: @BenChaimberg  - I have pushed a branch to demonstrate the issues with the request. I'll talk about the key points for `Binary` and `Token`:

`Binary`: IMHO I do not see a good use case for the support. I am interested if anybody can provide a good use case. I demonstrated that we could use [Typed Arrays](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray) to handle processing byte arrays and identifying them as Binary. I ran into a final issue converting those to strings and paused while this discussion happens (I think TextEncoder may be the answer). In order to use this function a developer would basically need to read in some binary data and shove it into the CloudFormation (it's constant at this point, plus or minus CFN condition magic). If I reached a situation where I needed to do that, I would put the object in S3 or Dynamo and use other means to access it. The more likely use case would be that some function dumps a chunk of binary data back into the state payload. This will either be a string array of size 1 or a base64 encoded string. I can't think of any way to introspect this is binary data during synth. For this reason I think if we must support `Binary` I would propose an optional attribute and moving back to a defined MessageAttribute Interface. Again, for the StepFunction use case IMO I think Binary is near obsolete.

`Token`: Will "work" in this current implementation due to all the amazing work that Rico's put into Token <-> String interops. However, I think that may prove problematic. The StepFunctions implementation as is *does not* support substitutions. I say that because of the work for [integrationResourceArn](https://github.com/aws/aws-cdk/blob/7966f8d48c4bff26beb22856d289f9d0c7e7081d/packages/%40aws-cdk/aws-stepfunctions-tasks/lib/private/task-utils.ts#L29-L35). Tokens are used pervasively in CDK, but some may expect a `Ref: <ResourceID>` to actually resolve and with this limitation it will silently fail until the user tests the SFN code via deploy. 

In summary I encourage CDK to not support Binary, but will disagree and commit by changing back to an interface. I am open to proposals on how best to handle `Token` and I did consider making it fail, but that felt too rigid.

**Comment by moofish32** [comment]
This comment https://github.com/aws/aws-cdk/pull/14817#discussion_r666534979 represents what I believe to be blocking finishing this work.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 15]
```diff
@@ -92,14 +104,72 @@ export class SnsPublish extends sfn.TaskStateBase {
...
+function handleMessageAttributeValue(attributeValue: any): messageAttributeValue {
+  if (attributeValue instanceof sfn.TaskInput) {
+    return { DataType: STRING, StringValue: attributeValue.value };
+  }
+  if (isByteArray(attributeValue)) {
+    return { DataType: BINARY, BinaryValue: '' };
+  }
+  if (Array.isArray(attributeValue)) {
+    // validates the primitives
+    attributeValue.forEach(v => handlePrimitiveValues(v));
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:151]: I think we may get into trouble here at some point if `handlePrimitiveValues` diverges from the spec of what we can put in this array (if we can make tokens work then we should be able to allow a token in this array). Let's create a separate validation function

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 16]
```diff
@@ -92,14 +104,72 @@ export class SnsPublish extends sfn.TaskStateBase {
...
+  BinaryValue?: string;
+};
+
+const STRING = 'String';
+const STRING_ARRAY = 'String.Array';
+const NUMBER = 'Number';
+const BINARY = 'Binary';
+
+function handleMessageAttributeValue(attributeValue: any): messageAttributeValue {
+  if (attributeValue instanceof sfn.TaskInput) {
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:143]:
```suggestion
  if (sfn.TaskInput.isEncodedJsonPath(attributeValue)) {
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/sns/publish.test.ts [hunk 17]
```diff
@@ -38,6 +38,63 @@ describe('Publish', () => {
...
+  test('with message attributes', () => {
+    // GIVEN
+    const stack = new cdk.Stack();
+    const topic = new sns.Topic(stack, 'Topic');
+
+    // WHEN
+    const task = new SnsPublish(stack, 'Publish', {
+      topic,
+      message: sfn.TaskInput.fromText('Publish this message'),
+      messageAttributes: {
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/test/sns/publish.test.ts:50]: add task input and token cases

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/sns/publish.test.ts [hunk 18]
```diff
@@ -38,6 +38,63 @@ describe('Publish', () => {
...
+        cake: 'chocolate',
+        cakePic: new Uint8Array(16),
+        cakeCount: 2,
+        vendors: ['Great Cakes', 'Cakes R Us', 'Local Cakes'],
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/test/sns/publish.test.ts:54]: add a boolean and null

**Review by BenChaimberg — CHANGES_REQUESTED** [review]
I like the auto-detect, feels very powerful for our customers! Just a few major comments, rest are naming/implementation details

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts (hunk 4 above)
Comment (moofish32) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts]: The issue is Token resolution, that part just works. The issue is the prior decision to not support [Step Function Substitution](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-stepfunctions-statemachine.html#cfn-stepfunctions-statemachine-definitionsubstitutions). 

Let's use an example with the current code, if I set up the following test case:

```ts
  test('with message attributes', () => {
    // GIVEN
    const stack = new cdk.Stack();
    const topic = new sns.Topic(stack, 'Topic');
    const stream = new Stream(stack, 'TokenExample');

    // WHEN
    const task = new SnsPublish(stack, 'Publish', {
      topic,
      message: sfn.TaskInput.fromText('Publish this message'),
      messageAttributes: {
        cake: 'chocolate',
        cakePic: new Uint8Array(16),
        cakeCount: 2,
        vendors: ['Great Cakes', 'Cakes R Us', 'Local Cakes'],
        stream: stream.streamArn,
      },
    });
```
The output render  the token for ARN correctly as:

```ts
 ... snip ...
     "stream": {
             "DataType": "String",
             "StringValue":  {
               "Fn::GetAtt":  [
                 "TokenExample182385A1",
                 "Arn",
               ],
             },
... snip ...
```

However because we don't define substitution correctly this either fails at deploy time or shoves that into a string and fails at runtime because the ARN doesn't actually get replaced correctly. The SFN definition substitution needs to get created and that needs to be `${SubstituteName}` instead of the CFN Fn::GetAtt function. If in fact the Token is just a string, then this just works. However, I feel like this gotcha will be a very bad user experience. A broader refactor that inspects for any `Token` in the StepFunction Definition and automatically sets the a substitution for it will work, but that's a much larger feature than this addition.

**Comment by moofish32** [comment]
I agree with comments @BenChaimberg and will get a refactor up this week.

**Review by BenChaimberg — DISMISSED** [review]
just trying to get this off my queue; please re-request my review when you're ready for me to take another look

**Comment by moofish32** [comment]
@BenChaimberg  - I think I've now addressed all of your comments. Can you please give me another review?

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 19]
```diff
@@ -98,8 +163,90 @@ export class SnsPublish extends sfn.TaskStateBase {
...
+    return { DataType: dataType, BinaryValue: `${attribute.value}` };
+  }
+
+  if (Token.isUnresolved(attribute.value)) {
+    return { DataType: dataType ?? MessageAttributeDataType.STRING, StringValue: attribute.value };
+  }
+
+  validateMessageAttribute(attribute);
+  if (Array.isArray(attribute.value)) {
+    return { DataType: dataType ?? MessageAttributeDataType.STRING_ARRAY, StringValue: JSON.stringify(attribute.value) };
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:204]:
```suggestion
    return { DataType: MessageAttributeDataType.STRING_ARRAY, StringValue: JSON.stringify(attribute.value) };
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 20]
```diff
@@ -98,8 +163,90 @@ export class SnsPublish extends sfn.TaskStateBase {
...
+    return;
+  }
+  if (Array.isArray(value)) {
+    if (dataType !== MessageAttributeDataType.STRING_ARRAY) {
+      throw new Error(`Unsupported SNS message attribute type: ${JSON.stringify(value)}`);
+    }
+    const validArrayTypes = ['string', 'boolean', 'number'];
+    value.forEach((v) => {
+      if (v !== null || !validArrayTypes.includes(typeof v)) {
+        throw new Error(`Unsupported SNS message attribute in array: ${v}`);
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:227]:
```suggestion
        throw new Error(`Unsupported SNS message attribute type ${typeof v} in array: ${v}`);
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 21]
```diff
@@ -98,8 +163,90 @@ export class SnsPublish extends sfn.TaskStateBase {
...
+
+function validateMessageAttribute(attribute: MessageAttribute): void {
+  const dataType = attribute.dataType;
+  const value = attribute.value;
+  if (dataType === undefined) {
+    return;
+  }
+  if (Array.isArray(value)) {
+    if (dataType !== MessageAttributeDataType.STRING_ARRAY) {
+      throw new Error(`Unsupported SNS message attribute type: ${JSON.stringify(value)}`);
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:222]: For this and the other error messages below, can we use the following pattern:
> Requested SNS message attribute type was ${dataType} but value was of type [array/string/...]

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 22]
```diff
@@ -98,8 +163,90 @@ export class SnsPublish extends sfn.TaskStateBase {
...
+  Object.entries(attributes).map(([key, val]) => {
+    renderedAttributes[key] = renderMessageAttributeValue(val);
+  });
+  return sfn.TaskInput.fromObject(renderedAttributes).value;
+}
+
+function renderMessageAttributeValue(attribute: MessageAttribute): MessageAttributeValue {
+  const dataType = attribute.dataType;
+  if (attribute.value instanceof sfn.TaskInput) {
+    return { DataType: dataType ?? MessageAttributeDataType.STRING, StringValue: attribute.value.value };
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:191]:
```suggestion
    return {
      DataType: dataType ?? MessageAttributeDataType.STRING,
      StringValue: dataType !== MessageAttributeDataType.BINARY ? attribute.value.value : undefined,
      BinaryValue: dataType === MessageAttributeDataType.BINARY ? attribute.value.value : undefined
    };
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 23]
```diff
@@ -98,8 +163,90 @@ export class SnsPublish extends sfn.TaskStateBase {
...
+        MessageAttributes: renderMessageAttributes(this.props.messageAttributes),
         Subject: this.props.subject,
       }),
     };
   }
 }
+
+interface MessageAttributeValue {
+  DataType: string;
+  StringValue?: string | null;
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:175]: unless I'm missing something, I don't think this is necessary
```suggestion
  StringValue?: string;
```

**Review by BenChaimberg — DISMISSED** [review]
This is looking great! Just a couple of minor things and I think we'll be good to go!

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts (hunk 21 above)
Comment (moofish32) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:226]: Is it worth trying to print the value? Right now the message will tell you the value doesn't match the type but no which value?

Comment (BenChaimberg) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts]: Yes – will add this!

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts [hunk 24]
```diff
@@ -98,8 +163,95 @@ export class SnsPublish extends sfn.TaskStateBase {
...
+
+function validateMessageAttribute(attribute: MessageAttribute): void {
+  const dataType = attribute.dataType;
+  const value = attribute.value;
+  if (dataType === undefined) {
+    return;
+  }
+  if (Array.isArray(value)) {
+    if (dataType !== MessageAttributeDataType.STRING_ARRAY) {
+      throw new Error(`Requested SNS message attribute type was ${dataType} but value was of type Array`);
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts:226]:
```suggestion
      throw new Error(`Requested SNS message attribute type was ${dataType} but ${value} was of type Array`);
```

**Review by BenChaimberg — CHANGES_REQUESTED** [review]

**Review by BenChaimberg — APPROVED** [review]
